cli/types/index.d.ts

```diff
@@ -174,757 +182,785 @@ declare namespace Cypress {
+    wait(ms: number, options?: Partial<Loggable & Timeoutable>): Chainable<undefined>
+    wait(alias: string, options?: Partial<Loggable & Timeoutable>): Chainable<WaitXHR>
+    wait(alias: string[], options?: Partial<Loggable & Timeoutable>): ChainableArray<WaitXHR>
 
     /**
      * Get the window object of the page that is currently active.
      *
-     * @param {Loggable} [options]
-     * @returns {Chainable}
-     * @memberof Chainable
      * @see https://on.cypress.io/window
-     * @example
-     *    cy.visit('http://localhost:8080/app')
-     *    cy.window().then(function(win){
-     *      // win is the remote window
-     *      // of the page at: http://localhost:8080/app
-     *    })
+     * @example window
+     * ```ts
+     * cy.visit('http://localhost:8080/app')
+     * cy.window().then(function(win){
+     *   // win is the remote window
+     *   // of the page at: http://localhost:8080/app
+     * })
+     * ```
      */
-    window(options?: Loggable): Chainable
+    window(options?: Partial<Loggable & Timeoutable>): Chainable<Window>
 
     /**
+     * Scopes all subsequent cy commands to within this element. Useful when working within a particular group of elements such as a `<form>`.
      * @see https://on.cypress.io/within
      */
-    within(fn: (currentSubject?: any) => void): Chainable
-    within(options: Loggable, fn: (currentSubject?: any) => void): Chainable // inconsistent argument order
+    within(fn: (currentSubject: Subject) => void): Chainable<Subject>
+    within(options: Partial<Loggable>, fn: (currentSubject?: Subject) => void): Chainable<Subject> // inconsistent argument order
 
     /**
+     * Yield the object passed into `.wrap()`.
+     *
      * @see https://on.cypress.io/wrap
      */
-    wrap(Object: object, options?: Loggable): Chainable
+    wrap<E extends Node = HTMLElement>(element: E | JQuery<E>, options?: Partial<Loggable & Timeoutable>): Chainable<JQuery<E>>
+    wrap<S>(objects: S[], options?: Partial<Loggable & Timeoutable>): ChainableArray<S>
+    wrap<S>(object: S, options?: Partial<Loggable & Timeoutable>): Chainable<S>
 
     /**
+     * Write to a file with the specified contents.
+     *
      * @see https://on.cypress.io/writefile
      */
-    writeFile(filePath: string, contents: FileContents, options?: Timeoutable): Chainable
-    writeFile(filePath: string, contents: FileContents, encoding: string, options?: Timeoutable): Chainable
+    writeFile<C extends FileContents>(filePath: string, contents: C, options?: Partial<Loggable>): Chainable<C>
+    writeFile<C extends FileContents>(filePath: string, contents: C, encoding: Encodings, options?: Partial<Loggable>): Chainable<C>
   }
 
-  interface DebugOptions {
-    verbose?: boolean
+  /**
+   * Chainable interface with stronger typing for array subjects
+   */
+  interface ChainableArray<Subject> extends Omit<Chainable<Subject[]>, 'and' | 'as' | 'each' | 'should' | 'then'> {

```

Comment:
I added `ChainableArray<Subject>` to properly capture types for `.each` and `.spread` and theoretically any Command that is meant to work on arrays. Without this, `.each` and `.spread` would be forced to resort to `any` for parameters, putting more type burden on the application author.

Reply:
The `Omit` part basically takes the whole `Chainable` interface and removes some keys ('and', 'as', etc in this case). I'm defining the `ChainableArray` interface to be the same as `Chainable` except a few methods are replaced with different type definitions. I could simply extend but then the wrong overrides would be chosen.